### PR TITLE
DeDir: Add new pallet `Decentralized Directory`

### PR DESCRIPTION
**Pull request by vatsa287**
A `v1` implementation of DeDir - Decentralised Directory.

- [x] Basic Functionalities - `create-registry`, `create-registry-entry`, `registry-entry-state-change`.
- [x] Change states to be in `enum` format. Custom states will be from `attributes`.
- [x] Identifier creation/management.
- [x] Delegation/ Registry-Ownership management.

Abandoned. Do not merge.

### CODE REVIEW

**Review by amarts — DISMISSED** [review]
LGTM

Is there any demo script planned to be implemented for the same?

**Comment by vatsa287** [comment]
> LGTM
> 
> Is there any demo script planned to be implemented for the same?

@amarts 
Sure it will be implemented.
It would be great if you can merge [this PR](https://github.com/dhiway/cord.js/pull/231), so I can push on top of this.

File: pallets/dedir/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,456 @@
...
+		///
+		/// # Example
+		/// ```
+		/// create_registry(origin, registry_id, attrs, additional_states)?;
+		/// ```
+		#[pallet::call_index(0)]
+		#[pallet::weight({0})]
+		pub fn create_registry(
+			origin: OriginFor<T>,
+			registry_id: RegistryIdOf<T>,
```
Comment (smohan-dw) [inline pallets/dedir/src/lib.rs]: Registry ID is an identifier that should be returned after a successful operation.

File: pallets/dedir/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,456 @@
...
+			attrs: Vec<(RegistryKeyIdOf<T>, RegistrySupportedTypeOf)>,
```
Comment (smohan-dw) [inline pallets/dedir/src/lib.rs]: The registry entry attributes should also help to identify the lookup fields that should be used for creating entry identifiers

File: pallets/dedir/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,456 @@
...
+			// TODO: Use Runtime Constant for upper bound
+			additional_states: Option<BoundedVec<BoundedVec<u8, ConstU32<32>>, ConstU32<10>>>,
```
Comment (smohan-dw) [inline pallets/dedir/src/lib.rs]: why can't this be enums? also why can't this be part of attrs?

File: pallets/dedir/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,456 @@
...
+			}
+
+			/* Ensure new-states does not overflow */
+			if let Some(states) = additional_states {
+				if registry.add_supported_states(states).is_err() {
+					return Err(Error::<T>::AddingStateFailed.into());
+				}
+			}
+
+			Registries::<T>::insert(&registry_id, registry);
```
Comment (smohan-dw) [inline pallets/dedir/src/lib.rs:327]: registry ownership?

File: pallets/dedir/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,456 @@
...
+		/// ```
+		/// create_registry_entry(origin, registry_id, registry_entry_id,
+		///     digest, attrs)?;
+		/// ```
+		#[pallet::call_index(1)]
+		#[pallet::weight({0})]
+		pub fn create_registry_entry(
+			origin: OriginFor<T>,
+			registry_id: RegistryIdOf<T>,
+			registry_entry_id: RegistryEntryIdOf<T>,
```
Comment (smohan-dw) [inline pallets/dedir/src/lib.rs]: again registry_entry_id is the result of the operation

File: pallets/dedir/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,456 @@
...
+			digest: EntryHashOf<T>,
+			attrs: Option<Vec<(RegistryKeyIdOf<T>, Data)>>,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
```
Comment (smohan-dw) [inline pallets/dedir/src/lib.rs]: need to use this to check the registry ownership or delegation

**Review by smohan-dw — CHANGES_REQUESTED** [review]

File: pallets/dedir/src/lib.rs (hunk 2 above)
Comment (vatsa287) [reply pallets/dedir/src/lib.rs]: With the discussion with @amarts  previously if the identifier is available then the registry_entry attributes would be fetch able.

File: pallets/dedir/src/lib.rs (hunk 3 above)
Comment (vatsa287) [reply pallets/dedir/src/lib.rs]: Ack.
As per newer design the states supported shall not be dynamic as implemented above but fixed to `Active`, `Revoked`, `Draft` which will become enum type.
So any custom state can be part of attrs itself.

File: pallets/dedir/src/lib.rs (hunk 1 above)
Comment (vatsa287) [reply pallets/dedir/src/lib.rs]: As per earlier conversation `identifier` creation was passed onto app side. 
Would move the logic here if required.

File: pallets/dedir/src/lib.rs (hunk 4 above)
Comment (vatsa287) [reply pallets/dedir/src/lib.rs:327]: Delegation/Ownership is not implemented in this version. Will do with next commit.

**Comment by amarts** [comment]
done with #504

**Pull request closed without merging**